src/adapters/env.ts

```diff
@@ -0,0 +1,110 @@
+import { z, ZodError } from "zod";
+
+import { StrictSchema } from "src/utils/type-safety";
+import { ZodEither } from "src/utils/types";
+// adapters
+import * as adapters from "src/adapters";
+
+interface ProductionLiteral {
+  REACT_APP_ENV: "production";
+}
+
+interface ProductionVars {
+  REACT_APP_INFURA_API_KEY: string;
+  REACT_APP_PRICE_UPDATER_API_URL: string;
+  REACT_APP_PRICE_UPDATER_API_KEY: string;
+  REACT_APP_WALLETCONNECT_BRIDGE: string;
+}
+
+type ProductionEnv = ProductionLiteral & ProductionVars;
+
+interface DevelopmentLiteral {
+  REACT_APP_ENV: "development";
+}
+
+type DevelopmentVars = ProductionVars & {
+  REACT_APP_HERMEZ_API_URL: string;
+  REACT_APP_HERMEZ_CONTRACT_ADDRESS: string;
+  REACT_APP_WITHDRAWAL_DELAYER_CONTRACT_ADDRESS: string;
+  REACT_APP_BATCH_EXPLORER_URL: string;
+  REACT_APP_ETHERSCAN_URL: string;
+};
+
+type DevelopmentEnv = DevelopmentLiteral & DevelopmentVars;
+
+type Env = ProductionEnv | DevelopmentEnv;
+
+const productionLiteralParser = StrictSchema<ProductionLiteral>()(
+  z.object({
+    REACT_APP_ENV: z.literal("production"),
+  })
+);
+
+const productionVarsParser = StrictSchema<ProductionVars>()(
+  z.object({
+    REACT_APP_INFURA_API_KEY: z.string(),
+    REACT_APP_PRICE_UPDATER_API_URL: z.string(),
+    REACT_APP_PRICE_UPDATER_API_KEY: z.string(),
+    REACT_APP_WALLETCONNECT_BRIDGE: z.string(),
+  })
+);
+
+const productionEnvParser = StrictSchema<ProductionEnv>()(
+  productionLiteralParser.and(productionVarsParser)
+);
+
+const developmentLiteralParser = StrictSchema<DevelopmentLiteral>()(
+  z.object({
+    REACT_APP_ENV: z.literal("development"),
+  })
+);
+
+const developmentVarsParser = StrictSchema<DevelopmentVars>()(
+  productionVarsParser.and(
+    z.object({
+      REACT_APP_HERMEZ_API_URL: z.string(),
+      REACT_APP_HERMEZ_CONTRACT_ADDRESS: z.string(),
+      REACT_APP_WITHDRAWAL_DELAYER_CONTRACT_ADDRESS: z.string(),
+      REACT_APP_BATCH_EXPLORER_URL: z.string(),
+      REACT_APP_ETHERSCAN_URL: z.string(),
+    })
+  )
+);
+
+const developmentEnvParser = StrictSchema<DevelopmentEnv>()(
+  developmentLiteralParser.and(developmentVarsParser)
+);
+
+export function getEnv(): ZodEither<Env, ZodError<Env>> {
```

Comment:
I thought that we agreed on not using monads!

Reply:
I would check if it's exported from `zod` then. I prefer to do that than creating our own monad type for this (and having it explicitly in our code)

Reply:
Well... In fact, we could only use `parse` instead of `safeParse`, at the end this function should only return the `Env` domain entity.

Reply:
I already checked and they don't. It's declared inline:

![image](https://user-images.githubusercontent.com/15896177/150514878-9216364c-78b2-45ed-9813-e286c107e5b6.png)

Reply:
Wait! they just added the type in the [last version](https://github.com/colinhacks/zod/pull/732/files)! Do you want me to upgrade to [3.11.6](https://github.com/colinhacks/zod/releases/tag/v3.11.6)?

Reply:
I think that this a better approach:

> Well... In fact, we could only use `parse` instead of `safeParse`, at the end this function should only return the Env domain entity.

The adapters should return the domain entities, the same as we are doing in `hermez-api` adapters etc

Reply:
We agreed in renaming the `ZodEither` to simply `Either`.

Reply:
Done